lib/Dialect/TTIR/Transforms/HoistCPUOps.cpp

```diff
@@ -0,0 +1,296 @@
+    mlir::Block *block = hoistedFunc.addEntryBlock();
+    mlir::OpBuilder builder(block, block->end());
+
+    // Map operands to block arguments and clone the operation
+    llvm::SmallVector<mlir::Value> newOperands;
+    for (auto operand : llvm::enumerate(opToHoist->getOperands())) {
+      newOperands.push_back(block->getArgument(operand.index()));
+    }
+
+    mlir::IRMapping mapping;
+    for (auto operand : llvm::zip(opToHoist->getOperands(), newOperands)) {
+      mapping.map(std::get<0>(operand), std::get<1>(operand));
+    }
+
+    mlir::Operation *clonedOp = builder.clone(*opToHoist, mapping);
+
+    // Add a return operation to the function
+    builder.create<mlir::func::ReturnOp>(opToHoist->getLoc(),
+                                         clonedOp->getResults());
+
+    // Declare the function prototype in the source module
+    localFunc =
+        func::FuncOp::create(opToHoist->getLoc(), functionName, funcType);
+    localFunc.setPrivate(); // Mark as external (no body)
+    sourceModule.push_back(localFunc);
+
+    hoistedFunc->setAttr("arg_ranks", builder.getI64ArrayAttr(ranks));
+  }
+
+  // Replace the original operation with a call to the hoisted function
+  mlir::OpBuilder opBuilder(opToHoist);
+  auto callOp = opBuilder.create<mlir::func::CallOp>(
+      opToHoist->getLoc(), localFunc, opToHoist->getOperands());
+
+  // Replace all results of the original operation with the call results
+  for (auto result : llvm::zip(opToHoist->getResults(), callOp.getResults())) {
+    std::get<0>(result).replaceAllUsesWith(std::get<1>(result));
+  }
+
+  // Erase the original operation
+  opToHoist->erase();
+}
+
+// an analysis class which relies on manual input locations to mark ops to be
+// hoisted, useful for testing
+class TTIRHoistAnalyzeManual {
+public:
+  using HoistOpSet = std::vector<std::set<mlir::Operation *>>;
+
+  TTIRHoistAnalyzeManual(mlir::Operation *op,
+                         const std::vector<std::string> &targetLocs) {
+    auto moduleOp = llvm::dyn_cast<mlir::ModuleOp>(op);
+    assert(moduleOp && "somehow got non-ModuleOp in TTIRHoistAnalyzeManual!");
+
+    moduleOp.walk([&](mlir::Operation *nestedOp) {
+      if (matchesLocation(nestedOp, targetLocs)) {
+        std::set<mlir::Operation *> opSet;
+        opSet.insert(nestedOp);
+        hoistedOps.push_back(opSet);
+      }
+    });
+  }
+
+  HoistOpSet getResults() { return hoistedOps; }
+
+private:
+  HoistOpSet hoistedOps;
+
+  static bool matchesLocation(mlir::Operation *op,
+                              const std::vector<std::string> &targetLocations) {
+    if (const auto nameLoc = dyn_cast<NameLoc>(op->getLoc())) {
+      const auto &opLocString = nameLoc.getName().str();
+
+      for (const auto &loc : targetLocations) {
+        if (opLocString == loc) {
+          return true;
+        }
+      }
+    }
+    return false;
+  }
+};
+
+struct TargetLocsParser : llvm::cl::parser<std::vector<std::string>> {
```

Comment:
Is this used implicitly by mlir? If so please add comment why.

Reply:
No, I should have removed this, sorry.  (I promise I'm usually better at self-reviewing before publishing PR, just made a lot of changes at once this time and am context switching between PRs a lot, haha).